kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/SchemaUtils.java

```diff
@@ -0,0 +1,375 @@
+    return new SchemaGenerator(config).inferIcebergType(value);
+  }
+
+  static class SchemaGenerator {
+
+    private int fieldId = 1;
+    private final IcebergSinkConfig config;
+
+    SchemaGenerator(IcebergSinkConfig config) {
+      this.config = config;
+    }
+
+    @SuppressWarnings("checkstyle:CyclomaticComplexity")
+    Type toIcebergType(Schema valueSchema) {
+      switch (valueSchema.type()) {
+        case BOOLEAN:
+          return BooleanType.get();
+        case BYTES:
+          if (valueSchema.name() != null && valueSchema.name().equals(Decimal.LOGICAL_NAME)) {
+            int scale = Integer.parseInt(valueSchema.parameters().get(Decimal.SCALE_FIELD));
+            return DecimalType.of(38, scale);
+          }
+          return BinaryType.get();
+        case INT8:
+        case INT16:
+          return IntegerType.get();
+        case INT32:
+          if (valueSchema.name() != null) {
+            if (valueSchema.name().equals(Date.LOGICAL_NAME)) {
+              return DateType.get();
+            } else if (valueSchema.name().equals(Time.LOGICAL_NAME)) {
+              return TimeType.get();
+            }
+          }
+          return IntegerType.get();
+        case INT64:
+          if (valueSchema.name() != null && valueSchema.name().equals(Timestamp.LOGICAL_NAME)) {
+            return TimestampType.withZone();
+          }
+          return LongType.get();
+        case FLOAT32:
+          return FloatType.get();
+        case FLOAT64:
+          return DoubleType.get();
+        case ARRAY:
+          Type elementType = toIcebergType(valueSchema.valueSchema());
+          if (config.schemaForceOptional() || valueSchema.valueSchema().isOptional()) {
+            return ListType.ofOptional(nextId(), elementType);
+          } else {
+            return ListType.ofRequired(nextId(), elementType);
+          }
+        case MAP:
+          Type keyType = toIcebergType(valueSchema.keySchema());
+          Type valueType = toIcebergType(valueSchema.valueSchema());
+          if (config.schemaForceOptional() || valueSchema.valueSchema().isOptional()) {
+            return MapType.ofOptional(nextId(), nextId(), keyType, valueType);
+          } else {
+            return MapType.ofRequired(nextId(), nextId(), keyType, valueType);
+          }
+        case STRUCT:
+          List<NestedField> structFields =
+              valueSchema.fields().stream()
+                  .map(
+                      field ->
+                          NestedField.of(
+                              nextId(),
+                              config.schemaForceOptional() || field.schema().isOptional(),
+                              field.name(),
+                              toIcebergType(field.schema())))
+                  .collect(toList());
+          return StructType.of(structFields);
+        case STRING:
+        default:
+          return StringType.get();
+      }
+    }
+
+    Optional<Type> inferIcebergType(Object value) {
```

Comment:
Style: Iceberg doesn't generally use `Optional` and just passes `null` instead.

Reply:
I changed these to do null checks instead.